zephyr/lib/regions_mm.c

```diff
@@ -0,0 +1,49 @@
+// SPDX-License-Identifier: BSD-3-Clause
+/*
+ * Copyright(c) 2022 Intel Corporation. All rights reserved.
+ *
+ * Author: Jakub Dabek <jakub.dabek@intel.com>
+ */
+
+#include <sof/lib/regions_mm.h>
+
+struct virtual_memory_heap
+	vm_heaps[CONFIG_MP_MAX_NUM_CPUS + VIRTUAL_REGION_COUNT] = { {0} };
+
+/**
+ * @brief Fill vm_heaps array with information from zephyr.
+ *
+ * Virtual memory regions calculated in zephyr are translated here
+ * to a struct that will keep all information on current allocations
+ * and virtual to physical mappings that are related to heaps.
+ * System heap is not a part of this information. It only refers to
+ * virtual first heaps.
+ * Has to be initialized after calculations for regions is done in zephyr.
+ */
+static int virtual_heaps_init(const struct device *unused)
+{
+	ARG_UNUSED(unused);
+
+	struct sys_mm_drv_region *virtual_memory_regions = { 0 };
+
+	virtual_memory_regions = (struct sys_mm_drv_region *)sys_mm_drv_query_memory_regions();
+
+	for (size_t i = 0;
+		i < CONFIG_MP_MAX_NUM_CPUS + VIRTUAL_REGION_COUNT;
+		i++) {
+		vm_heaps[i].virtual_region = &virtual_memory_regions[i];
+		if (virtual_memory_regions[i].attr & MEM_REG_ATTR_CORE_HEAP)
+			vm_heaps[i].memory_caps =
+				SOF_MEM_CAPS_RAM | SOF_MEM_CAPS_HP | SOF_MEM_CAPS_CACHE;
+
+		if (virtual_memory_regions[i].attr & MEM_REG_ATTR_SHARED_HEAP)
```

Comment:
You're overwriting `vm_heaps[i].memory_caps` in each of these `if`s, right? Can `.attr` have more than one of these flags set? If so, it seems that `MEM_REG_ATTR_OPPORTUNISTIC_MEMORY` has the highest priority, since it overrides the other two, and `MEM_REG_ATTR_CORE_HEAP` has the lowest priority. So you revert the order and use `else if` to make the priorities clearer.

Reply:
Each of those regions is a distinct entity so none of these has a priority perse. Zephyr provides information and i translate it here so we can use it further in allocation.
Those Attributes are work in progress - they are due to change if future implementation uses theme. However elseif is a good point.

Reply:
but you should clearly define (1) whether those attributes are mutually exclusive and (2) what are their priorities. The way you had it implemented in the previous version meant, that they possibly weren't mutually exclusive and the priorities were (from the highest to the lowest):
```
MEM_REG_ATTR_OPPORTUNISTIC_MEMORY
MEM_REG_ATTR_SHARED_HEAP
MEM_REG_ATTR_CORE_HEAP
```
In your today's version you exactly reverted priorities. Is this intended? OTOH, if these attributes are mutually exclusive, then priorities don't matter, then I would suggest to do
```
switch (virtual_memory_regions[i].attr & MEM_REG_ATTR_MASK) {
case MEM_REG_ATTR_SHARED_HEAP:
...
```

Reply:
Yea they are actually mutually exclusive and there are no priorities - need to change zephyr implementation to enum later or is there any other way You suggest ?